vm/src/builtins/tuple.rs

```diff
@@ -56,11 +56,30 @@ impl IntoPyTuple for Vec<PyObjectRef> {
     }
 }
 
-macro_rules! impl_into_pyobj_tuple {
-    ($(($T:ident, $idx:tt)),+) => {
+pub trait FromPyTuple<'a>: Sized {
+    fn from_pytuple(tuple: &'a PyTuple, vm: &VirtualMachine) -> PyResult<Self>;
+}
+
+macro_rules! impl_from_into_pytuple {
+    ($($T:ident),+) => {
         impl<$($T: ToPyObject),*> IntoPyTuple for ($($T,)*) {
             fn into_pytuple(self, vm: &VirtualMachine) -> PyTupleRef {
-                PyTuple::new_ref(vec![$(self.$idx.to_pyobject(vm)),*], &vm.ctx)
+                #[allow(non_snake_case)]
+                let ($($T,)*) = self;
+                PyTuple::new_ref(vec![$($T.to_pyobject(vm)),*], &vm.ctx)
+            }
+        }
+
+        // TODO: figure out a way to let PyObjectRef implement TryFromBorrowedObject, and
```

Comment:
`PyObjectRef` cannot, but `PyObject` can. Since `PyObject` can always be borrowed from `PyObjectRef`, I expect it has same effect.

```rust
impl<'a> TryFromBorrowedObject<'a> for &'a PyObject {
    fn try_from_borrowed_object(_vm: &VirtualMachine, obj: &'a PyObject) -> PyResult<Self> {
        Ok(obj)
    }
}
```

Reply:
Could you explain more about it?
For me, it doesn't seem possible without increasing refcount. And increasing a refcount is something not expected from TryFromBorrowedObject

Reply:
Oh, I didn't realize that was part of the contract for TryFromBorrowedObject. It seems useful to me to have some way of generalizing over a type that can be extracted from a borrowed pyobj, but doesn't necessarily *have* to borrow from it - like serde's Deserialize vs DeserializeOwned. Like this case - the T in extract_tuple<T>() should be *able* to borrow from the pyobjectref in the tuple, but there's no reason we should *enforce* it - a `(PyObjectRef,)` is a perfectly reasonable thing to put there.

Reply:
Thanks, I think I am going to catch up.